src/thor/bidirectional_astar.cc

```diff
@@ -351,6 +355,10 @@ bool BidirectionalAStar::ExpandReverse(GraphReader& graphreader,
   if (!costing_->Allowed(nodeinfo)) {
     const DirectedEdge* opp_edge = nullptr;
     const GraphId opp_edge_id = graphreader.GetOpposingEdgeId(pred.edgeid(), opp_edge, tile);
+    // Mark the predecessor as a deadend to be consistent with how the
+    // edgelabels are set when an *actual* deadend (i.e. some dangling OSM geometry)
+    // is labelled
+    pred.set_deadend(true);
```

Comment:
@kevinkreiser I'm not sure my tests hit the reverse case here, so a sanity check would be good: is `pred` in the reverse test supposed to be marked as a dead-end?  I assume it's the directed edge heading *away* from the deadend point?

The logic in `AutoCost::AllowedReverse()` simply looks at `pred.deadend()`, which wasn't the inverted logic I was expecting.

Is `deadend` supposed to be set on *both* directed edges connected to a deadend?  If yes, then all is well here, but it's a bit confusing as you can't tell without further digging whether an edge is coming *from* a deadend or going *to* one.

Reply:
Same question for `timedep_reverse` I guess - I copy-pasta'd the one-line "fix", but is that actually what we want?

Reply:
`DirectedEdge::deadend` is deprecated, we should note that in its javadoc comment. when we made the original change to the algorithms to find derived deadends we considered marking it spare but we couldnt because then old code using new tiles wouldnt work properly. so we have to keep properly setting that flag/bit at tile creation but the current algorithms shouldnt use it they should derive deadend-ness as you and dan and anders and i have been adding to the algorithms.

an easy way to test that its doing what you want in reverse is to set and arrive by time and put the destination behind a uturn at a node. this will trigger timedep_reverse which should work given your changes. the reverse search uses the opposing edges of the real edges we use on the path, thats why those methods take an opposing edge for allowed so it can check the real edge the path would be taking

Reply:
@kevinkreiser Hmm, not quite deprecated.  `DirectedEdge::deadend()` is still used in the constructor of `EdgeLabel` (I tried removing the getter function from DirectedEdge and it leads to compilation failure) where labels inherit whatever value is set on the directededge.  I will look and see if that is easy to remove so we can fully deprecate it to avoid future confusion.

Reply:
yeah it was many months ago, i remember looking at it with @PureW but i couldnt remember the outcome exactly. its probably better to leave it for now until dijkstras (and the matrix algorithms and multimodal and bss) get derived deadend support anyway.

Reply:
@kevinkreiser ok, for this PR I'll leave it - but boy, there sure is a lot of logic duplication in here.  I would expect a lot of this stuff to either live in a base class, or all these algorithms to derive from a template interface that supplied much of this kind of detail.

Reply:
ive eye'd that up too a couple of times but it just doesnt work out for slightly annoying reasons/differences between the algorithms. there might be a way to slice it just right so the code can be shared but it might involve using the same edge labels everywhere or some tricks with the astar heuristic. its definitely tempting but the devils in the details :smile:

Reply:
Well, the downside to *not* doing it is that the implementations in these algorithms are slowly diverging over time, leading to inconsistent results between actions (e.g. many_to_many between A->B could return different results to a route from A->B).